### ISSUE TITLE
LSTM-MESSL combination 4

### ISSUE THREAD

**Issue by mim** [speechLabBcCuny/messlJsalt15#7]
Train an LSTM that takes as input both a noisy speech signal and an estimated mask (MESSL output) and predicts a better mask.  Run it on the final mask estimated by MESSL to produce a cleaner final mask.

**Comment by nateanl** [comment]
I'm not clear about the procedure. So you mean I predict the mask by the new LSTM model. Then apply the predicted mask as the initial for the MESSL, just once? Or can we do multiple times?

**Comment by mim** [comment]
I think you meant to make that comment on #6.  This issue is about an LSTM that takes noisy speech and a rough mask as input and produces a cleaner mask.

**Comment by nateanl** [comment]
I mean after we get a cleaner mask by LSTM and send it to MESSL. Will we use the final one as the input for LSTM again, and predict a new mask, and send it to MESSL. We can do many times.

**Comment by mim** [comment]
There is no sending the cleaner mask back to MESSL here.  That is #8.  In this one, it's a one-shot process.  Noisy speech + rough mask in, clean mask out, use that clean mask to reconstruct the audio.

**Comment by nateanl** [comment]
Got it.

**Comment by grezesf** [comment]
How do we get enough rough masks to train the model?

**Comment by mim** [comment]
From the current MESSL outputs.

**Comment by mim** [comment]
The MESSL masks (or any masks used as input to a DNN) should be transformed with a logit function first (this is in the [mimlib repo](https://github.com/mim/mimlib/blob/master/logit.m)):

```matlab
function y = logit(x)

x = max(0, min(1, x));
y = log(x) - log(1-x);
```

**Comment by grezesf** [comment]
also will write a short doc with description of saved models and where to find them

**Comment by grezesf** [comment]
1. Why must the logit function be applied?
2. Where are the MESSL masks on the server?

**Comment by mim** [comment]
Because the mask is between 0 and 1, if you look at a histogram of mask values, it tends to have a U-shape, which is difficult for the DNN to use as input (well, slows down learning).  If you put it through the logit, then it turns into more of two gaussian-ish bumps, which is easier to learn.

The MESSL masks are here: `/scratch/mim/chime3/messlMcMvdrMrf.2Hard5Lbp4Slate/`.  The masks themselves are saved in the `data` subdirectory.  You'll have to look through those mat files to see which variable it is.

**Comment by grezesf** [comment]
there is a 0 byte file that probably needs to be deleted:
`F04_422C0211_BUS_2.mat ` in `/scratch/mim/chime3/messlMcMvdrMrf.2Hard5Lbp4Slate/data/dt05_bus_real`

**Comment by grezesf** [comment]
expriment running, log and models saved in: `/scratch/felix/IS2017-CHIME3-Exps/exp-LSTM-Mask-Cleaner`. will be moved to Saved-Experiments when it finishes, with a short text description.
code of the experiment is on the `messlkeras` branch (yay for trying to use git properly) in `keras_exp2_runner_mask_cleaner.py`.

some potential problems:
- logit puts values to be between 0 and 1, but also applies log(x) and log(1-x) which aren't defined when x=0 or 1. I min/maxed to 10^-50

**Comment by mim** [comment]
10^-50 seems quite small.  Does that affect a lot of points?  If a lot of points (more than 20%) are exactly at 0 or 1, then don't bother with the logit.

**Comment by nateanl** [comment]
It seems like we need to remove this file: 
`scratch/mim/chime3/messlMcMvdrMrf.2Hard5Lbp4Slate/data/dt05_bus_real\F04_422C0211_BUS_2.mat`

**Comment by mim** [comment]
I've deleted it.


On Mar 4, 2017 4:33 PM, "Zhaoheng Ni" <notifications@github.com> wrote:

It seems like we need to remove this file:
scratch/mim/chime3/messlMcMvdrMrf.2Hard5Lbp4Slate/data/dt05_bus_
real\F04_422C0211_BUS_2.mat

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/speechLabBcCuny/messlJsalt15/issues/7#issuecomment-284184327>,
or mute the thread
